Use Docker Compose v2, wait for healthchecks to pass

robertknight opened the pull request:
Switch to Docker Compose v2 and change `make services` to use the `--wait` flag to wait for healthchecks to pass before exiting. Also add a healthcheck for Elasticsearch. The Makefile changes follow the pattern in the lms app.

Together these changes mean that h and its services can be started with `make services && make dev`, without encountering errors due to Postgres / Elasticsearch not being ready to connect.

The `start_period` setting is set to a generous value (1 min) as normal startup times for ES are quite long (15-20s on my 2020 Intel MacBook Pro).

See also https://github.com/hypothesis/lms/pull/5556.

Example output from `make services`:

```
$ make services
[+] Running 3/3
 ✔ Container h-elasticsearch-1  Healthy                                           24.9s
 ✔ Container h-postgres-1       Healthy                                            2.4s
 ✔ Container h-rabbit-1         Healthy                                            1.9s
```

The startup time for Elasticsearch, as measured by whether `curl http://localhost:9200` returns a non-200 response, seems rather long to me, and also somewhat variable (~17s seems more typical). It might be possible to optimize that, but it is out of scope for this PR.

seanh commented on docker-compose.yml:
```diff
@@ -13,6 +13,10 @@ services:
     image: hypothesis/elasticsearch:latest
     ports:
       - '127.0.0.1:9200:9200'
+    healthcheck:
+        test: curl --fail --silent http://localhost:9200 >/dev/null
+        interval: 3s
+        start_period: 1m
```
Is this necessary? Does h actually fail to start up if Elasticsearch is not running yet? I know `make dev` will actually crash if Postgres isn't up yet, but I'm not sure whether the same is true for Elasticsearch and RabbitMQ. If we can avoid having to wait for Elasticsearch and RabbitMQ then we can keep things simpler (no healthchecks) and save a few seconds on the startup time

robertknight replied:
> Does h actually fail to start up if Elasticsearch is not running yet?

h tries to initialize the Elasticsearch index when it starts, so you get some confusing error messages if it isn't running. Also things will just seem broken if you try to access various pages served by h during the time when Elasticsearch is starting. I think it would be better to direct efforts into speeding up Elasticsearch startup time, as it seems to me that it really ought to be possible for it to start up quicker than it currently does.

seanh reviewed (changes requested):
h doesn't use the cookiecutter (yet) so we can edit these files directly in h :+1:

Can you also deleted the `requirements/dockercompose.{in,txt}` files?

Also these:

* https://github.com/hypothesis/h/blob/d08ae62859d7e2e940a7b9e6365e201b407e2edb/tox.ini#L88
* https://github.com/hypothesis/h/blob/d08ae62859d7e2e940a7b9e6365e201b407e2edb/tox.ini#L110
* https://github.com/hypothesis/h/blob/d08ae62859d7e2e940a7b9e6365e201b407e2edb/Makefile#L29
* https://github.com/hypothesis/h/blob/d08ae62859d7e2e940a7b9e6365e201b407e2edb/docs/developing/troubleshooting.rst?plain=1#L21

robertknight commented:
> Can you also deleted the requirements/dockercompose.{in,txt} files?

Done. I've updated / removed the various references to `docker-compose` and the dockercompose tox env.
